include/re_types.h

```diff
@@ -270,3 +270,76 @@ typedef size_t re_sock_t;
+	int:			sizeof(int),                                  \
+	unsigned int:		sizeof(unsigned int),                         \
+	long:			sizeof(long),                                 \
+	unsigned long:		sizeof(unsigned long),                        \
+	long long:		sizeof(long long),                            \
+	unsigned long long:	sizeof(unsigned long long),                   \
+	float:			sizeof(float),                                \
+	double:			sizeof(double),                               \
+	char const*:		sizeof(char const *),                         \
+	char*:			sizeof(char *),                               \
+	void const*:		sizeof(void const *),                         \
+	void*:			sizeof(void *),                               \
+	default: sizeof(void*)                                                \
+)
+
+#define RE_ARG_0() 0
+#define RE_ARG_1(expr) RE_ARG_SIZE(expr), (expr), 0
+#define RE_ARG_2(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_1(__VA_ARGS__)
+#define RE_ARG_3(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_2(__VA_ARGS__)
+#define RE_ARG_4(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_3(__VA_ARGS__)
+#define RE_ARG_5(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_4(__VA_ARGS__)
+#define RE_ARG_6(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_5(__VA_ARGS__)
+#define RE_ARG_7(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_6(__VA_ARGS__)
+#define RE_ARG_8(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_7(__VA_ARGS__)
+#define RE_ARG_9(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_8(__VA_ARGS__)
+#define RE_ARG_10(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_9(__VA_ARGS__)
+#define RE_ARG_11(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_10(__VA_ARGS__)
+#define RE_ARG_12(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_11(__VA_ARGS__)
+#define RE_ARG_13(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_12(__VA_ARGS__)
+#define RE_ARG_14(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_13(__VA_ARGS__)
+#define RE_ARG_15(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_14(__VA_ARGS__)
+#define RE_ARG_16(expr, ...) RE_ARG_SIZE(expr), (expr), RE_ARG_15(__VA_ARGS__)
+
+#define RE_ARG_VA_NUM_2(X, X16, X15, X14, X13, X12, X11, X10, X9, X8, X7, X6, \
+			X5, X4, X3, X2, X1, N, ...)                           \
+	N
+#define RE_ARG_VA_NUM(...)                                                    \
+	RE_ARG_VA_NUM_2(0, ##__VA_ARGS__, 16, 15, 14, 13, 12, 11, \
+			10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0)
+
+#define RE_ARG_N3(N, ...) RE_ARG_##N(__VA_ARGS__)
+#define RE_ARG_N2(N, ...) RE_ARG_N3(N, __VA_ARGS__)
+#define RE_VA_ARGS(...) RE_ARG_N2(RE_ARG_VA_NUM(__VA_ARGS__), __VA_ARGS__)
+#endif
+
+#define RE_VA_ARG(ap, val, type, safe)                                        \
+	if ((safe)) {                                                         \
+		size_t sz = va_arg((ap), size_t);                             \
+		assert(sz && "RE_VA_ARG: no more arguments");                 \
+		assert(sz <= sizeof(type) &&                                  \
```

Comment:
Should we consider a warning instead of assert ?

Reply:
It would be nice, if a test fails if size is mismatched or no more args available.
cmake sets `-DNDEBUG`, on `CMAKE_BUILD_TYPE=Release`  builds and `assert()` is ignored.

Reply:
Another approach is to return an error code if the format is wrong.

This means that e.g. mbuf_printf would return the error code.

Reply:
> Another approach is to return an error code if the format is wrong.

Most printf code ignores the return code I think, but its already set and exited here:
```c
		if (!sz || sz > sizeof(type)) {                               \
			err = EINVAL;                                         \
			goto out;                                             \
		}               
```
With Release build type assert is not called (I can make this more visible with `#ifndef RELEASE`)